packages/gatsby-cli/src/reporter/loggers/ink/cli.js

```diff
@@ -51,6 +53,30 @@ class CLI extends React.Component {
       )
     }
 
+    /*
+      Only operation on messages array is to push new message into it. Once
+      message is there it can't change. Because of that we can do single
+      transform from message object to react element and store it.
+      This will avoid calling React.createElement completely for every message
+      that can't change.
+    */
+    if (messages.length > this.memoizedReactElementsForMessages.length) {
+      for (
+        let index = this.memoizedReactElementsForMessages.length;
+        index < messages.length;
+        index++
+      ) {
+        const msg = messages[index]
+        this.memoizedReactElementsForMessages.push(
```

Comment:
Is React.createElement that expensive? With `React.memo` we shouldn't rerender these elements.

Reply:
Yes, `React.memo` avoids re-rendering, but `createElement` is still super expensive (at least for now, while we use development react version).

With just React.memo my "stress test" https://github.com/pieh/ink-stress-test still takes 5s to log 1000 messages. With memoized stuff it take 1.5s

On the left - with `memoizedReactElementsForMessages` and memo, on the right just with memo

![Screenshot 2020-03-04 at 15 28 45](https://user-images.githubusercontent.com/419821/75889421-32615400-5e2d-11ea-9902-4b0b5eff6311.png)
![Screenshot 2020-03-04 at 15 31 10](https://user-images.githubusercontent.com/419821/75889423-342b1780-5e2d-11ea-992b-5e9a8ba181ef.png)

I can upload .cpuprofile files if you want to take closer look ;)

Reply:
Notice on flamegraph on the right how much work our `message.map` is taking compared to `finishSyncRender` which "commit" changes to CLI output - the `finishSyncRender` in general is constant here - just the the `render` function of CLI component is massively more expensinve because of `React.createElement` (well at least for now - maybe swapping to production react build would make it a lot better, but this needs more engineering work - need webpack or rollup to bundle that stuff up and setup package build and hopefully watch correctly - so this is just effort vs impact calculation here ;) )

Reply:
I also do realize that doing what I'm doing is not really the "react" way, but it's pragmatic quick solution to a problem.

There are additional things we could explore, but those might need reaching out to `ink` maintainer about allowing `<Static>` component to not have to keep getting all past static elements - that way we could avoid mapping over all past messages on each re-render etc - but this seems even more involved and return of investment (time/research) might be questionable

Reply:
Wow that's harsh, 🔥 createElement 🔥

Reply:
Note, that this might be greatly impacted by using dev version of react which does a lot more validation stuff than production one (tho I didn't research that enough to give impact estimates of switching to prod react version - still want to do it - it's just more work over doing this hack)